drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt

```diff
@@ -8,191 +8,150 @@
 import android.widget.ImageView
 import android.widget.TextView
 import androidx.recyclerview.widget.RecyclerView
 import com.adyen.checkout.base.api.ImageLoader
-import com.adyen.checkout.base.model.paymentmethods.PaymentMethod
-import com.adyen.checkout.base.model.paymentmethods.StoredPaymentMethod
-import com.adyen.checkout.base.util.DateUtils
 import com.adyen.checkout.base.util.PaymentMethodTypes
+import com.adyen.checkout.core.exception.CheckoutException
 import com.adyen.checkout.core.log.LogUtil
 import com.adyen.checkout.dropin.R
 
 @SuppressWarnings("ComplexMethod", "TooManyFunctions")
 class PaymentMethodAdapter(
-    private var paymentMethodsModel: PaymentMethodsModel,
-    private var imageLoader: ImageLoader,
-    private var showInExpandStatus: Boolean,
+    private val paymentMethodsListModel: PaymentMethodsListModel,
+    private val imageLoader: ImageLoader,
     private val onPaymentMethodSelectedCallback: OnPaymentMethodSelectedCallback
 ) : RecyclerView.Adapter<PaymentMethodAdapter.BaseViewHolder>() {
 
-    private var expandListOfItem = emptyList<PaymentMethod>()
-    private var collapseListOfItem = emptyList<PaymentMethod>()
-
-    init {
-        initAdapter()
-    }
-
-    companion object {
-        internal val TAG = LogUtil.getTag()
-
-        internal const val PAYMENT_METHODS_HEADER = 1
-        internal const val PAYMENT_METHODS_FOOTER = 2
-        internal const val PAYMENT_METHOD_ITEMS = 3
-
-        internal val HEADER_PLACEHOLDER_ITEM_SP = PaymentMethod()
-        internal val HEADER_PLACEHOLDER_ITEM_P = PaymentMethod()
-        internal val FOOTER_PLACEHOLDER_ITEM = PaymentMethod()
-
-        private const val CARD_LOGO_TYPE = "card"
-    }
+    private val headerCount =
+        if (paymentMethodsListModel.storedPaymentMethods.isEmpty()) 1
+        else 2
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
         return when (viewType) {
             PAYMENT_METHODS_HEADER -> PaymentMethodsHeaderVH(getView(parent, R.layout.payment_methods_list_header))
-            PAYMENT_METHODS_FOOTER -> PaymentMethodsFooterVH(getView(parent, R.layout.payment_methods_list_footer))
-            else -> PaymentMethodVH(getView(parent, R.layout.payment_methods_list_item))
+            STORED_PAYMENT_METHOD -> StoredPaymentMethodVH(getView(parent, R.layout.payment_methods_list_item))
+            PAYMENT_METHOD -> PaymentMethodVH(getView(parent, R.layout.payment_methods_list_item))
+            else -> throw CheckoutException("Unexpected viewType on onCreateViewHolder - $viewType")
         }
     }
 
     override fun getItemViewType(position: Int): Int {
-        return when (getPaymentMethod(position)) {
-            HEADER_PLACEHOLDER_ITEM_P, HEADER_PLACEHOLDER_ITEM_SP -> PAYMENT_METHODS_HEADER
-            FOOTER_PLACEHOLDER_ITEM -> PAYMENT_METHODS_FOOTER
-            else -> PAYMENT_METHOD_ITEMS
+        // Top is always a header
+        return if (position == 0) {
```

Comment:
We can replace this with one `when` block?

```
    override fun getItemViewType(position: Int): Int {
        // Top is always a header
        return when {
            (position == 0) -> PAYMENT_METHODS_HEADER
            noStored() -> PAYMENT_METHOD // only regular payment methods left
            position <= paymentMethodsListModel.storedPaymentMethods.size -> STORED_PAYMENT_METHOD // 0 header - stored payments
            position == (paymentMethodsListModel.storedPaymentMethods.size + 1) -> PAYMENT_METHODS_HEADER //2nd header - regular payments
            else -> PAYMENT_METHOD
        }
    }
```

Reply:
👍